docs/src/main/sphinx/security/opa-access-control.md

```diff
@@ -0,0 +1,265 @@
+- `resource`: information about the accessed objects
+- `targetResource`: information about any newly created object, if applicable
+- `grantee`: grantee of a grant operation.
+
+Fields that are not applicable for a specific operation are set to null.
+Examples are an empty  `targetResource` if not modifying a table or schema or
+catalog is modified, or an empty `grantee` if not granting permissions is set.
+Any null field is omitted altogether from the `action` object.
+
+### Example requests to OPA
+
+Accessing a table results in a query similar to the following example:
+
+```json
+{
+  "context": {
+    "identity": {
+      "user": "foo",
+      "groups": ["some-group"]
+    },
+    "softwareStack": {
+      "trinoVersion": "434"
+    }
+  },
+  "action": {
+    "operation": "SelectFromColumns",
+    "resource": {
+      "table": {
+        "catalogName": "example_catalog",
+        "schemaName": "example_schema",
+        "tableName": "example_table",
+        "columns": [
+          "column1",
+          "column2",
+          "column3"
+        ]
+      }
+    }
+  }
+}
+```
+
+The `targetResource` is used in cases where a new resource, distinct from the one in
+`resource` is created. For example, when renaming a table.
+
+```json
+{
+  "context": {
+    "identity": {
+      "user": "foo",
+      "groups": ["some-group"]
+    },
+    "softwareStack": {
+      "trinoVersion": "434"
+    }
+  },
+  "action": {
+    "operation": "RenameTable",
+    "resource": {
+      "table": {
+        "catalogName": "example_catalog",
+        "schemaName": "example_schema",
+        "tableName": "example_table"
+      }
+    },
+    "targetResource": {
+      "table": {
+        "catalogName": "example_catalog",
+        "schemaName": "example_schema",
+        "tableName": "new_table_name"
+      }
+    }
+  }
+}
+```
+
+(opa-batch-mode)=
+## Batch mode
+
+A very powerful feature provided by OPA is its ability to respond to
+authorization queries with more complex answers than a `true` or `false` boolean
+value.
+
+Many features in Trino require filtering to determine to which resources a user
+is granted access. These resources are catalogs, schema, queries, views, and
+others objects.
+
+If `opa.policy.batched-uri` is not configured, Trino sends one request to OPA
+for each object, and then creates a filtered list of accessible objects.
+response was returned.
+
+Configuring `opa.policy.batched-uri` allows Trino to send a request to
+the batch endpoint, with a list of resources in one request using the
+under `action.filterResources` node.
+
+All other fields in the request are identical to the non-batch endpoint.
+
+An OPA policy supporting batch operations should return a list containing the
+_indices_ of the items for which authorization is granted. Returning a `null`
+value or an empty list is equivalent and denies any access.
+
+You can batching support for policies that't originally do not support it:
+
+```text
+package foo
```

Comment:
As reported in Slack https://trinodb.slack.com/archives/C04HV35BJFR/p1704222429750449
```suggestion
package foo

import future.keywords.contains
```

Reply:
Updated